From b4273917645e9b3fca5385648f8056652a07070b Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 2 May 2014 14:36:38 -0400
Subject: [misc] mei: mei_write correct checks for copy_from_user

Message-id: <1399041470-21330-6-git-send-email-prarit@redhat.com>
Patchwork-id: 79294
O-Subject: [RHEL7.1 PATCH BZ 1088953 05/77] mei: mei_write correct checks for copy_from_user
Bugzilla: 1088953
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1088953

commit d8b29efa27de17dafe516d45204dbb744ccdfa48
Author: Alexander Usyskin <alexander.usyskin@intel.com>
Date:   Mon Sep 2 03:11:02 2013 +0300

    mei: mei_write correct checks for copy_from_user

    1. return -EFUALT when copy_from_user fails
    2. display error message on failure in error level

    Signed-off-by: Alexander Usyskin <alexander.usyskin@intel.com>
    Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/misc/mei/main.c b/drivers/misc/mei/main.c
index 0c29f2b..dcdcbf5 100644
--- a/drivers/misc/mei/main.c
+++ b/drivers/misc/mei/main.c
@@ -414,8 +414,11 @@ static ssize_t mei_write(struct file *file, const char __user *ubuf,
   goto out;
 
  rets = copy_from_user(write_cb->request_buffer.data, ubuf, length);
- if (rets)
+ if (rets) {
+  dev_err(&dev->pdev->dev, "failed to copy data from userland\n");
+  rets = -EFAULT;
   goto out;
+ }
 
  cl->sm_state = 0;
  if (length == 4 &&
@@ -587,7 +590,7 @@ static long mei_ioctl(struct file *file, unsigned int cmd, unsigned long data)
  dev_dbg(&dev->pdev->dev, "copy connect data from user\n");
  if (copy_from_user(connect_data, (char __user *)data,
     sizeof(struct mei_connect_client_data))) {
-  dev_dbg(&dev->pdev->dev, "failed to copy data from userland\n");
+  dev_err(&dev->pdev->dev, "failed to copy data from userland\n");
   rets = -EFAULT;
   goto out;
  }
-- 
1.7.1